### PR TITLE
Custom UEFI certificates - API change

### PR DESCRIPTION
**Pull request by benjamreis**
Solves: https://github.com/xapi-project/xen-api/issues/5245

### CODE REVIEW

File: ocaml/idl/datamodel_pool.ml
Comment (lindig) [inline ocaml/idl/datamodel_pool.ml:1033]: would drop "used"

File: ocaml/xapi/xapi_globs.ml
Comment (lindig) [inline ocaml/xapi/xapi_globs.ml:1449]: .. setting *a* custom location ..

File: ocaml/xapi/xapi_pool.ml
Comment (lindig) [inline ocaml/xapi/xapi_pool.ml:3573]: This suggests the empty string has a special meaning. Is this obvious for the user?

Comment (benjamreis) [reply ocaml/xapi/xapi_pool.ml:3573]: In the design clearing the custom UEFI certificates would default back to using the default ones.

Comment (psafont) [reply ocaml/xapi/xapi_pool.ml:3573]: Some docs can be added clarifying this clearing. An alternative could be to add an api function called `Pool.clear_custom_uefi_certs`

File: ocaml/idl/datamodel_pool.ml
Comment (psafont) [inline ocaml/idl/datamodel_pool.ml:1029]: Leave this empty so it can get autogenerated

File: ocaml/idl/datamodel_pool.ml
Comment (psafont) [inline ocaml/idl/datamodel_pool.ml:1409]: Same here

File: ocaml/xapi/xapi_pool.ml
Comment (stormi) [reply ocaml/xapi/xapi_pool.ml:3573]: Note that this special meaning of an empty value is already present currently in `Db.Pool.{set,}uefi_certificates`. We just moved this to the new field, to keep the amount of changes to a minimum.

Comment (benjamreis) [reply ocaml/xapi/xapi_pool.ml:3573]: I've added in the doc of the setter the defaulting back when empty behavior. :)

**Comment by benjamreis** [comment]
All comments should have addressed and we have tested this PR against our CI so this should be go for review and hopefully merge.
I'll resolve the schematest just before merge as it might change again.

**Comment by robhoes** [comment]
If you have a field called `pool.uefi_certificates` then a getter called `pool.get_uefi_certificates` will always be generated, unless the field is marked as `internal_only` (which may have other implications).

**Comment by benjamreis** [comment]
It means it's not possible to override `pool.get_uefi_certificates`?

**Comment by robhoes** [comment]
> It means it's not possible to override pool.get_uefi_certificates?

If there is a public field and you change its getter, then that could lead to inconsistent behaviour. For example, the value of the field will also be returned by calls like `pool.get_record` and `pool.get_all_records`, and in events returned by `event.from`, but that would be the value as in the database.

**Comment by stormi** [comment]
> > It means it's not possible to override pool.get_uefi_certificates?
> 
> If there is a public field and you change its getter, then that could lead to inconsistent behaviour. For example, the value of the field will also be returned by calls like `pool.get_record` and `pool.get_all_records`, and in events returned by `event.from`, but that would be the value as in the database.

So the issue is not technical and it's mostly consistency of the API? As long as the behaviour of the getter is documented, the override looks good to me, because it offers a single API call to get the actually used certs, without causing any changes on XenServer side. It didn't raise any objections at the design stage in #5245. However, if this is seen as a bad thing, then we'll need to simply forget about this "one API call gives the used certs" thing and just query both custom then default certificates from the client side.

> I don't fully understand the design here, but why not change the value of the actual `pool.uefi_certificates` field in the DB with the custom value, depending on the configuration?

This is what the previous implementation did, but this caused several issues detailed on top of #5245. Two fields is cleaner.

**Comment by benjamreis** [comment]
We removed the override of `Pool.get_uefi_certificates` and we'll let the xapi client fetch the custom ones or the default ones whe necessary.

File: ocaml/xapi/message_forwarding.ml
Comment (psafont) [inline ocaml/xapi/message_forwarding.ml:1103]: Just in case, the certificates printed here don't usually contain private keys, do they?

File: ocaml/xapi/xapi_host.ml
Comment (psafont) [inline ocaml/xapi/xapi_host.ml:2728]: these names are outdated, using `debug "%s: ..." __FUNCTION__ ...` would be better

**Review by psafont — COMMENTED** [review]
I can't spot any big issues, just a doubt around printing certificates, and the reminder to update the versions in the datamodel when this is ready

**Comment by psafont** [comment]
Just remembered, ocaml/xapi-cli-server/records.ml needs to be modified so xe can push the certificates

**Comment by stormi** [comment]
> Just remembered, ocaml/xapi-cli-server/records.ml needs to be modified so xe can push the certificates

Regarding this, maybe we could wait for the need to arise from users. We already have a script which downloads certificates from the net then sets them directly through a XAPI call, and also allows setting custom certs.

I'm not against the principle, but time flies and we need this to move forward with user testing of XCP-ng 8.3 beta :)

File: ocaml/idl/datamodel_pool.ml
Comment (benjamreis) [reply ocaml/idl/datamodel_pool.ml:1032]: I'll update it right before the merge since it can chang again.

**Review by psafont — APPROVED** [review]
This simplifies the call graph when setting the custom uefi certificates (no more Host.set_uefi_certificates) I think it's worth adding the extra field and function calls to Pool here, at the cost of special handling in xapi_vm.

I searched for more users of uefi_certificates and couldn't find it, so the solution seems complete. If there were more users maybe a function in xapi_helpers could be added to handle this centrally. In the meantime this looks good.

Please do not merge yet, it needs changing versions in the datamodel

**Comment by benjamreis** [comment]
Which version should I put? `v23.32.0`?

File: ocaml/idl/datamodel_pool.ml
Comment (psafont) [inline ocaml/idl/datamodel_pool.ml:1046]: Please leave this lifecycle empty, I believe it should work

Comment (benjamreis) [reply ocaml/idl/datamodel_pool.ml]: Done!

**Review by psafont — APPROVED** [review]

**Review by psafont — APPROVED** [review]

**Comment by benjamreis** [comment]
last force push was to fix typos in the commit messages, everything is good and validated on our side so good for much according to us! :)

**Review by stormi — APPROVED** [review]

**Comment by benjamreis** [comment]
The format error does not concern my changes, not sure what to do :thinking:

**Comment by benjamreis** [comment]
rebased on master for format change.

**Comment by stormi** [comment]
How about a merge of this as a Christmas gift? :wink:

File: ocaml/idl/datamodel_pool.ml
Comment (lindig) [inline ocaml/idl/datamodel_pool.ml:1048]: typo? "verts"?

**Review by lindig — APPROVED** [review]

File: ocaml/idl/datamodel_pool.ml
Comment (benjamreis) [reply ocaml/idl/datamodel_pool.ml]: indeed! Fixed :innocent:

File: ocaml/idl/datamodel_pool.ml
Comment (psafont) [inline ocaml/idl/datamodel_pool.ml:1032]: This is the latest version

**Review by psafont — APPROVED** [review]

**Pull request merged**
